Doc: Campus Example Update (#5223)
aristanetworks/avd: ansible_collections/arista/avd/examples/campus-fabric/README.md

Reviewer: I assume this should now refer to SITE1_SPINE1.cfg instead? There could be other artifacts like that. Please ask AI to do naming health check. =)

Reviewer: Not related to this PR directly, but let's use it to update to Ma1 for cEOS-lab.

---
title: Campus Fabric # This title is used for search results
link: https://avd.arista.com/stable/ansible_collections/arista/avd/examples/campus-fabric/index.html
---
<!--
  ~ Copyright (c) 2025 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

> # Campus Fabric

Reviewer 2: link line 15 is broken and points to campus instead of L2LS

---
title: Campus Fabric # This title is used for search results
link: https://avd.arista.com/stable/ansible_collections/arista/avd/examples/campus-fabric/index.html
---
<!--
  ~ Copyright (c) 2025 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Campus Fabric

## Introduction

> This example describes and includes all the AVD files used to build a Campus Fabric by building upon the previous [L2LS Fabric](../../examples/l2ls-fabric/README.md) example. The spine nodes provide L3 routing of SVIs, and the L2 leaf nodes support 802.1x Network Access Control (NAC) with port ranges.

Reviewer: I'd remove "previous" and rework this a bit. My understanding is that campus will be a standalone inventory that we are going to maintain separately. In that case I'd simply mention that campus is the modification of L2LS and some knobs were already covered in the corresponding guide, but avoid explicitly mentioning that it's built on top.

Author: I like the shout-out here because it can lead the reader to the alternate guide if they have not covered it yet.

  - IDF2 supporting 240 users with a Modular five slot chassis (48 ports per module)
  - IDF3 supporting 480 users with five leafs (2RU - 96 ports each)

> The drawing below shows the physical topology used in this example. The interface assignments shown here are referenced across the entire example, so keep that in mind if this example is adapted to a different topology.

Reviewer: Make this a very explicit and visible warning instead somewhere on the top, stating that all interfaces and other vars must be adjusted accordingly when crafting an inventory for a different lab/prod network.

Author: It makes sense to place it here as it is the first mention of the topology layout.

Reviewer: No a critical remark. Up to you. The key point - if we want to highlight that, it's easier to have a universal warning for all examples in the future and avoid writing about changing specific vars. It's not about topology layout - it's about example being an example. One universal statement for everything in every example will work better.
We can move that into a separate PR for examples standardization.

> The drawing below shows the physical topology used in this example. The interface assignment shown here are referenced across the entire example, so keep that in mind if this example must be adapted to a different topology.

Reviewer 3: ```suggestion
The drawing below shows the physical topology used in this example. The interface assignments shown here are referenced across the entire example, so keep that in mind if this example is adapted to a different topology.
```

``` text
- ACME
  - S1
    - S1_SPINES
    - S1_LEAFS
  - NETWORK_SERVICES
    - S1_SPINES
    - S1_LEAFS
  - CONNECTED_ENDPOINTS
    - S1_SPINES
    - S1_LEAFS
```

ACME represents the highest level within the hierarchy. Ansible variables defined at this level will be applied to all nodes in the fabric. Ansible groups have parent-and-child relationships. For example, both S1_SPINES and S1_LEAFS are children of S1. Groups of Groups are possible and allow variables to be shared at any level within the hierarchy. For example, NETWORK_SERVICES is a group with two other groups defined as children: S1_SPINES and S1_LEAFS. The same applies to the group named CONNECTED_ENDPOINTS. You will see these groups listed at the bottom of the inventory file.

This naming convention makes it possible to extend anything quickly and can be changed based on your preferences. The names of all groups and hosts must be unique.

![Figure: 2](images/ansible_groups.svg)

### inventory.yml

> The below inventory file represents two spines and ten leafs. The nodes are defined under the groups S1_SPINES and S1_LEAFS, respectively. We apply group variables (group_vars) to these groups to define their functionality and configurations.

Reviewer 3: I think there's only 8 leafs, not 10 right?

Author: Correct, fixing now

> The below inventory file represents two spines and eight leafs. The nodes are defined under the groups SITE1_SPINES and SITE1_LEAFS, respectively. We apply group variables (group_vars) to these groups to define their functionality and configurations.

Reviewer: ```suggestion
The below inventory file represents a leaf-spine topology. The nodes are defined under the groups SITE1_SPINES and SITE1_LEAFS, respectively. We apply group variables (group_vars) to these groups to define their functionality and configurations.
```

## Management Network

> This example configures a dedicated management network on port Management0 (vrf: MGMT) and an in-band management network using SVI VLAN 10 (vrf: default). In-band management is easily configured with two variables under the leaf defaults key in `SITE1_LEAFS.yml`. First, it auto-generates an SVI and default route on each leaf node. Then, on the Spine nodes, it will build a matching SVI for VLAN 10 and create a Virtual-IP (10.10.10.1) for the defined subnet.

Reviewer: This is not what the example inventory shows. 1) Example is using l2leaf.defaults, not leaf.defaults. 2) under l2leaf.defaults platform is set to cEOS and management interface is Ma1, not Ma0

???+ note

>     The Management0 interface is only present in cEOS-lab variants of EOS.

Reviewer: This is not true any more since my last comment. While it is still the default for cEOS-lab, I don't think we can reference it here as AVD default for cEOS-lab is now Ma1.

## Management Network

> This example configures a dedicated management network on port Management0 (vrf: MGMT) and an in-band management network using SVI VLAN 10 (vrf: default). In-band management is easily configured with two variables under the leaf defaults key in `SITE1_LEAFS.yml`. First, it auto-generates an SVI and default route on each leaf node. Then, on the Spine nodes, it will build a matching SVI for VLAN 10 and create a Virtual-IP (10.10.10.1) for the defined subnet.

Reviewer: Not sure if it's listed somewhere, but if not - add a warning that Management0 is a port present on cEOS-lab only and will be different on every hardware switch. It kind of obvious for us, but not necessarily for every customer.

| LEAF3D | 172.16.100.109 | 10.10.10.12 |
| LEAF3E | 172.16.100.110 | 10.10.10.13 |

> In Campus Networks, having a dedicated out-of-band management network in each IDF is uncommon. Therefore, you can easily disable configuring the Management0 interface and the management VRF by adding these variables to the `SITE1_LEAFS.yml` group_vars.

Reviewer: change to generic "management interface" instead of "Management0 interface"

| NETWORK_SERVICES.yml      | VLANs/SVIs                                    |
| CONNECTED_ENDPOINTS.yml   | Port Profiles and Network Port Ranges         |

The tabs below show the Ansible **group_vars** used in this example.

=== "ACME"
    At the top level (ACME), the following variables are defined in the **group_vars/ACME.yml** file. These Ansible variables apply to all nodes in the fabric and is a common place to set AAA, users, NTP, and management interface settings. Update local_users and passwords for your environment.

    You can create a sha512_password by creating a username and password on a switch and copy/paste it within double quotes here.

    ``` yaml
    --8<--
    ansible_collections/arista/avd/examples/campus-fabric/group_vars/ACME.yml
    --8<--
    ```

=== "S1"
    At the Fabric level (S1), the following variables are defined in the **group_vars/S1.yml** file. The fabric name, underlay routing protocol, and the core interfaces model to connect our site to the WAN.

    ``` yaml
    --8<--
    ansible_collections/arista/avd/examples/campus-fabric/group_vars/S1.yml
    --8<--
    ```

=== "S1_SPINES"

>     In an L2LS Campus design, there are two types of spine nodes: `spine` and `l3spine`. For a spine node to provide routing of SVIs, we set the type to `l3spine`. Variables applied under the node key type (spine/leaf) defaults section are inherited by nodes under each type. These variables may be overwritten under the node itself.

Reviewer 3: ```suggestion
    In an L2LS Campus design, there are two types of spine nodes: `spine` and `l3spine`. For a spine node to provide routing of SVIs, we set the type to `l3spine`. Variables applied under the node key type (l3spine/l2leaf) defaults section are inherited by nodes under each type. These variables may be overridden under the node itself.
````

    ``` yaml
    --8<--
    ansible_collections/arista/avd/examples/campus-fabric/group_vars/S1_SPINES.yml
    --8<--
    ```

=== "S1_LEAFS"

>     In an L2LS Campus design, we have one type of leaf node: `l2leaf`. Variables applied under the node key type (spine/leaf) defaults section are inherited by nodes under each type. These variables may be overwritten under the node itself.

Reviewer 3: ```suggestion
    In an L2LS Campus design, we have one type of leaf node: `l2leaf`. Variables applied under the node key type (l3spine/l2leaf) defaults section are inherited by nodes under each type. These variables may be overridden under the node itself.
```

    ``` yaml
    --8<--
    ansible_collections/arista/avd/examples/campus-fabric/group_vars/NETWORK_SERVICES.yml
    --8<--
    ```

=== "CONNECTED_ENDPOINTS"
    Our fabric would not be complete without connecting some devices to it. Therefore, we define port profiles and network port ranges in the  **group_vars/CONNECTED_ENDPOINTS.yml** file. A single port_profile may be used across several switches and port ranges. In our example, we create a port profile called `PP-DOT1X` to define generic 802.1x (NAC) settings we wish to apply to a range of ports. The `network_ports` data model defines which switches and ports to apply the port profile. This data model allows a single regex statement to define a list of switches. In addition, the variable `switch_ports` expands into a range of ports. You can view more details of the `range_expand` filter [here](../../../../../docs/plugins/Filter_plugins/range_expand.md). These variables are applied to the spine and leaf nodes since they are a part of this inventory group.

    ``` yaml
    --8<--
    ansible_collections/arista/avd/examples/campus-fabric/group_vars/CONNECTED_ENDPOINTS.yml
    --8<--
    ```

## Network Services

> The Network Services data model is stored in the **NETWORK_SERVICES** group_var tab above. Each IDF will have three unique VLANs to support Data, Voice, and Guest networks. The spine nodes will provide routing for these VLANs via locally assigned SVIs.

Reviewer 3: ```suggestion
The Network Services data model is stored in the **NETWORK_SERVICES** group_vars tab above. Each IDF will have three unique VLANs to support Data, Voice, and Guest networks. The spine nodes will provide routing for these VLANs via locally assigned SVIs.
```

## Port Profiles and Network Ports

AVD provides a way to standardize and reuse port profiles through a compact data model you can utilize across the network. The Network Ports data model is stored in the **NETWORK_PORTS** group_vars tab above. Each port is configured to support NAC and dynamically assigns the proper VLAN based on 802.1x authentication. Multiple device types (IP Phones, Workstations, Printers, Access Points, etc.) can share the same port configuration below.

> ![Figure: 3](images/dot1x_ports.svg#only-light)

Reviewer: Consider using gray or similar background instead and switching to a single diagram. Maintaining light and dark diagrams at scale adds to maintenance and repository weight. In this specific case transparent background could be fine, appropriate font color that works across 2 color schemes could be fine. Usually background is a safer option though.

Author: In my experience, without two diagrams for light and dark mode, the diagram only looks good for one setting. Creating a font that works across both is very hard, especially when sticking to style guides.

Author: I think this is a good example of what you mean, one diagram that can work with both themes https://avd.arista.com/5.5/docs/getting-started/avd-aap.html#topology

Reviewer: yes, that is a good example. I can bring more examples if required. We have to reduce the number of diagrams as long as we handcraft them.

![Figure: 3](images/dot1x_ports_dark.svg#only-dark)

The above sample port configuration is easily produced with `port_profiles` and `network_ports` data models. Each port has similar configuration items defined in `port_profiles`, while `network_ports` defines which switches and port ranges are to be applied. The `network_ports` data model allows regex to match switches and an `expand_range` filter to cover a range of ports. For details, see the documentation for [`port_profiles`](../../roles/eos_designs/docs/input-variables.md#port-profiles-settings) and [`network_ports`](../../roles/eos_designs/docs/input-variables.md#network-ports-settings).

## WAN/Core Edge

> Lastly, we need to provide a way for traffic to exit the Campus Fabric via the WAN/Core Edge. This is easily accomplished using `underlay_routing_protocol` and `core_interfaces` to create point-to-point IP links to the WAN/Core and enable a routing protocol such as OSPF. This simple data model below is located at the bottom of the **group_vars/ACME_FABRIC.yml** file.

Reviewer 2: the file name seems wrong as we are pointing to SITE1.yml. I think the last sentecne should be removed.

After the playbook run finishes, EOS CLI intended configuration files were written to `intended/configs`.

> To build and deploy the configurations to your switches, run the playbook called `deploy.yml`. This assumes that your Ansible host has access and authentication rights to the switches. In this example, we defined the authentication variables in the `ACME.yml` file.

Reviewer: The AI shared a very valid concern that `deploy.yml` is not building configs any more. Also AI shared interesting point of view that I never though about: "That is a behavior regression for anyone expecting fresh configs from changed inputs before deployment; they will push whatever was already sitting under intended/." I still think having a dedicated deploy shortcut is best, but I'd ask AI to rewrite this sentence a bit based on the concerns above.

Author: Update playbook to reflect this section
